kombu/transport/cherami.py

```diff
@@ -0,0 +1,151 @@
+"""Cherami Transport."""
+
+from cherami_client.consumer import Consumer
+from cherami_client.publisher import Publisher
+
+from kombu.async import get_event_loop
+from kombu.log import get_logger
+from kombu.utils.json import loads, dumps
+
+from . import virtual
+
+
+DEFAULT_PORT = 4922
+
+logger = get_logger('kombu.transport.cherami')
+
+class Channel(virtual.Channel):
+    """Cherami Channel."""
+
+    default_message_batch_size = 1  # No. of messages to fetch when get message is called
+
+    _publisher = None
+    _consumer = None
+
+    # message fetching limit
+    prefetch_limit = 0
+    prefetched = 0
+
+    def __init__(self, *args, **kwargs):
+        super(Channel, self).__init__(*args, **kwargs)
+
+        # event_loop
+        self.hub = kwargs.get('hub') or get_event_loop()
+
+        # cherami-client instance kwargs
+        self.kwargs = self.connection.kwargs
+
+        # delivery_tag(Kombu.Transport.QoS) -> delivery_token(cherami)
+        self.delivery_map = dict()
+
+    def basic_consume(self, queue, no_ack, *args, **kwargs):
+        # set the prefetch_limit specified by the celery configuration
+        self.prefetch_limit = self.qos.prefetch_count
+
+        if self.hub:
+            self.hub.call_soon(self._get, queue)
+        # no_ack is True because cherami is in charge of message ack
+        return super(Channel, self).basic_consume(
+            queue, True, *args, **kwargs
+        )
+
+    def _put(self, queue, message, **kwargs):
+        # the first argument is the message id, cherami doesn't care about this, just passes along
+        self.publisher.publish(str(0), dumps(message))
+
+    def _get(self, queue, callback=None):
+        if self.prefetched < self.prefetch_limit:
+            try:
+                results = self.consumer.receive(num_msgs=self.default_message_batch_size)
+                self._on_message_ready(queue, results)
+            except Exception as e:
+                logger.info('Failed to receive messages: {0}'.format(e))
+        else:
+            self.hub.call_soon(self._get, queue)
+
+    def _on_message_ready(self, queue, results):
+        for res in results:
+            delivery_token = res[0]
```

Comment:
suggest renaming delivery_token to cherami_delivery_token, to avoid the confusion with delivery_tag